sqlglot/jsonpath.py

```diff
@@ -0,0 +1,210 @@
+        pass
+
+    def _match(token_type, raise_unmatched=False):
+        if _curr() == token_type:
+            return _advance()
+        if raise_unmatched:
+            raise ParseError(_error(f"Expected {token_type}"))
+        return None
+
+    def _parse_literal() -> t.Any:
+        token = _match(TokenType.STRING) or _match(TokenType.IDENTIFIER)
+        if token:
+            return token.text
+        if _match(TokenType.STAR):
+            return _node("wildcard")
+        if _match(TokenType.PLACEHOLDER):
+            start = i
+
+            while True:
+                if _match(TokenType.L_BRACKET):
+                    _parse_bracket()  # nested call which we can throw away
+                if _curr() in (TokenType.R_BRACKET, None):
+                    break
+                _advance()
+            return _node("filter", path[tokens[start].start : tokens[i].end])
+
+        number = "-" if _match(TokenType.DASH) else ""
+
+        token = _match(TokenType.NUMBER)
+        if token:
+            number += token.text
+
+        if number:
+            return int(number)
+        return False
+
+    def _parse_slice() -> t.Any:
+        start = _parse_literal()
+        end = _parse_literal() if _match(TokenType.COLON) else None
+        step = _parse_literal() if _match(TokenType.COLON) else None
+
+        if end is None and step is None:
+            return start
+        return _node("slice", start=start, end=end, step=step)
+
+    def _parse_bracket() -> JSONPathNode:
+        literal = _parse_slice()
+
+        if isinstance(literal, str) or literal is not False:
+            indexes = [literal]
+            while _match(TokenType.COMMA):
+                literal = _parse_slice()
+
+                if literal:
+                    indexes.append(literal)
+
+            if len(indexes) == 1:
+                if isinstance(literal, str):
+                    node = _node("key", indexes[0])
+                elif isinstance(literal, dict) and literal["kind"] == "filter":
+                    node = _node("selector", indexes[0])
+                else:
+                    node = _node("subscript", indexes[0])
+            else:
+                node = _node("union", indexes)
+        else:
+            raise ParseError(_error("Cannot have empty segment"))
+
+        _match(TokenType.R_BRACKET, raise_unmatched=True)
+
+        return node
+
+    nodes = []
+
+    while _curr():
+        if _match(TokenType.DOLLAR):
+            nodes.append(_node("root"))
+        elif _match(TokenType.DOT):
+            recursive = _prev().text == ".."
+            value = _match(TokenType.VAR) or _match(TokenType.STAR)
+            nodes.append(
+                _node("recursive" if recursive else "child", value=value.text if value else None)
+            )
+        elif _match(TokenType.L_BRACKET):
+            nodes.append(_parse_bracket())
+        elif _match(TokenType.VAR):
+            nodes.append(_node("key", _prev().text))
```

Comment:
What case does this cover? When can you come across a `VAR` outside of the `DOT` branch? Also, ditto about other tokens like `SELECT` that aren't `VAR` but could still be valid keys. They won't be recognized as `VAR`s.

Reply:
everything is a var basically, again, select and other special sql tokens don''t exist